programs/zstdcli_trace.c

```diff
@@ -0,0 +1,192 @@
+/* We only use macros from threading.h so it is compatible with dynamic linking */
+#include "../lib/common/threading.h"
+
+#if ZSTD_TRACE
+
+static FILE* g_traceFile = NULL;
+static int g_mutexInit = 0;
+static ZSTD_pthread_mutex_t g_mutex;
+
+void TRACE_enable(char const* filename)
+{
+    int const writeHeader = !UTIL_isRegularFile(filename);
+    if (g_traceFile)
+        fclose(g_traceFile);
+    g_traceFile = fopen(filename, "a");
+    if (g_traceFile && writeHeader) {
+        /* Fields:
+        * algorithm
+        * version
+        * method
+        * streaming
+        * level
+        * workers
+        * dictionary size
+        * uncompressed size
+        * compressed size
+        * duration nanos
+        * compression ratio
+        * speed MB/s
+        */
+        fprintf(g_traceFile, "Algorithm, Version, Method, Mode, Level, Workers, Dictionary Size, Uncompressed Size, Compressed Size, Duration Nanos, Compression Ratio, Speed MB/s\n");
+    }
+    if (!g_mutexInit) {
+        if (!ZSTD_pthread_mutex_init(&g_mutex, NULL)) {
+            g_mutexInit = 1;
+        } else {
+            TRACE_finish();
+        }
+    }
+}
+
+void TRACE_finish(void)
+{
+    if (g_traceFile) {
+        fclose(g_traceFile);
+    }
+    g_traceFile = NULL;
+    if (g_mutexInit) {
+        ZSTD_pthread_mutex_destroy(&g_mutex);
+        g_mutexInit = 0;
+    }
+}
+
+static void TRACE_log(char const* method, PTime duration, ZSTD_trace const* trace)
+{
+    int level = 0;
+    int workers = 0;
+    double const ratio = (double)trace->uncompressedSize / (double)trace->compressedSize;
+    double const speed = ((double)trace->uncompressedSize * 1000) / (double)duration;
+    if (trace->params) {
+        ZSTD_CCtxParams_getParameter(trace->params, ZSTD_c_compressionLevel, &level);
+        ZSTD_CCtxParams_getParameter(trace->params, ZSTD_c_nbWorkers, &workers);
+    }
+    assert(g_traceFile != NULL);
+
+    /* Fields:
+     * algorithm
+     * version
+     * method
+     * streaming
+     * level
+     * workers
+     * dictionary size
+     * uncompressed size
+     * compressed size
+     * duration nanos
+     * compression ratio
+     * speed MB/s
+     */
+    fprintf(g_traceFile,
+        "zstd, %u, %s, %s, %d, %d, %llu, %llu, %llu, %llu, %.2f, %.2f\n",
+        trace->version,
+        method,
+        trace->streaming ? "streaming" : "single-pass",
+        level,
+        workers,
+        (unsigned long long)trace->dictionarySize,
+        (unsigned long long)trace->uncompressedSize,
+        (unsigned long long)trace->compressedSize,
+        (unsigned long long)duration,
+        ratio,
+        speed);
+}
+
+static ZSTD_CCtx const* g_cctx;
+static ZSTD_DCtx const* g_dctx;
+static UTIL_time_t g_begin = UTIL_TIME_INITIALIZER;
+
+/**
+ * These symbols override the weak symbols provided by the library.
+ */
+
+int ZSTD_trace_compress_begin(ZSTD_CCtx const* cctx)
+{
+    int enabled = 0;
+    if (g_traceFile == NULL)
```

Comment:
Does this design mandates the use of global variables ?
Could it be mitigated by allowing passing some `void*` state ?

Reply:
There are two options:
1. Global state inside the tracing library.
2. return `void*` from the `begin()` function, and consume it in the `end()` function.

I chose the former, because the `void*` would have to be `malloc`'d. That would mean we have to be certain never to leak the `void*`. Currently `end()` is only called on success, so we would leak it on errors. We could fix that by calling `free()` whenever we overwrite it with a new `void*` or free the context. But, that introduces extra complexity to both zstd and the tracing library. Because now, what if the tracing library is in `C++`. It couldn't use `new` to allocate the object, it would have to use `malloc`.

So I decided to keep it simpler and have the tracing library keep track of the state. A more complex implementation would have a `map<uintptr_t, State>` to keep track of a separate state per-context, using the context pointer as a key.

Reply:
We could also change the zstd implementation to make it compulsory to call `end()`. Basically, if zstd detected that the `void* tracingContext` when calling `begin()` or during `ZSTD_freeCCtx()` it would call `end()` and set the `error` member to some generic error (and we would have the flexibility to set a more specific error code later).

Reply:
I understand the argument about simplicity.

It's just that, as it's a user API (even a hidden one), I'm easily worried that it could be more difficult to update it afterwards.

But I'm fine with the idea of starting simple, experiment, and introduce complexity only when it feels more palatable.

Reply:
> It's just that, as it's a user API (even a hidden one), I'm easily worried that it could be more difficult to update it afterwards.

At this point I expect this API to change between versions. Once we have enough experience with it, we can think about what stability guarantees we want to provide.